admin/app/controllers/solidus_admin/products_controller.rb

```diff
@@ -4,9 +4,50 @@ module SolidusAdmin
   class ProductsController < SolidusAdmin::BaseController
     def index
       set_page_and_extract_portion_from(
-        Spree::Product.order(created_at: :desc),
+        Spree::Product.order(created_at: :desc, id: :desc),
         per_page: SolidusAdmin::Config[:products_per_page]
       )
     end
+
+    def destroy
+      @products = Spree::Product.where(id: params[:id])
+
+      Spree::Product.transaction do
+        @products.discard_all
+      end
+
+      flash[:notice] = t('.success')
+      redirect_to products_path, status: :see_other
+    end
+
+    def discontinue
+      @products = Spree::Product.where(id: params[:id])
+
+      Spree::Product.transaction do
+        @products
+          .update_all(discontinue_on: Time.current)
+      end
+
+      flash[:notice] = t('.success')
+      redirect_to products_path, status: :see_other
+    end
+
+    def activate
+      @products = Spree::Product.where(id: params[:id])
+
+      Spree::Product.transaction do
```

Comment:
I'd like the logic in this controller to be somewhere else so it can be reused (for instance, by a worker). We don't have that place now, so let's move forward. :+1: